### PR TITLE
Api error handling

### PR DESCRIPTION
**Pull request by Keelando**
Implemented some error handling of the API responses and errors occurred from parsing the JSON response. Added scrolling to lattice view if rendered successfully

### CODE REVIEW

**Comment by Keelando** [comment]
Started tinkering with the error handling from the API.

The first errors propagate when the API returns nothing, which will generate an error in the AXIOS request if no response is received. This I believe would be a timeout. The errors I get from that are Network Error, and I'm not fully sure where that is originating.

the 2nd case errors are when the API returns a response != 200, such as CORS, or some other HTTP error.

The third case errors are when the API returns a response (200), but encountered some problem with compiling either due to empty file upload or something else. I encountered Type Error: Expected a string or bytes like object.

Ideally, I would like to only handle errors based on the response object of the AXIOS call, and based on this response, we would know whether JSON.parse(response.data) will succeed.

There is a subtlety there which we can discuss some time.

File: ls-ui-react-app/src/pages/UploadCircuitPage.tsx [hunk 1]
```diff
@@ -207,10 +222,51 @@ const UploadCircuitPage = ( {appState, setAppState} : AppStateProps)  =>
...
-                    { appState.compilationIsLoading && <b> Loading... </b> }
+                    { appState.compilationIsLoading && 
+                        <div>
+                            <div className="spinner-border text-success" role="status"></div>
+                            <b>Processing...</b>
+                        </div>
+                    }
+
+                    {/* If a request has been made, and compilation has finished, and compilation is still undefined, indicate error message */}
+                    { (appState.request === true && appState.compilationIsLoading === false && appState.compilationResult === undefined) &&
```
Comment (gwwatkin) [inline ls-ui-react-app/src/pages/UploadCircuitPage.tsx:235]:
```suggestion
                    { (appState.request  && !appState.compilationIsLoading && !appState.compilationResult ) && 
```

File: ls-ui-react-app/src/pages/UploadCircuitPage.tsx [hunk 2]
```diff
@@ -207,10 +222,51 @@ const UploadCircuitPage = ( {appState, setAppState} : AppStateProps)  =>
...
+                                </div>
+                            }
+
+                            {/* Response 200, but Compiler Error */}
+                            { appState.apiResponse.data && 
+                                <div className="alert alert-danger">
+                                    {appState.apiResponse.data.errorType + ": " + appState.apiResponse.data.errorMessage}
+                                </div>
+                            }
+
```
Comment (gwwatkin) [inline ls-ui-react-app/src/pages/UploadCircuitPage.tsx:256]: This is some very clear handling of error cases I can see everything that's happening

File: ls-ui-react-app/src/appState.ts [hunk 3]
```diff
@@ -8,6 +8,8 @@ class AppState {
     compilationResult? : CompilationResult
     errorMsg? : string // An optional error message to show to the user
     compilationIsLoading : boolean = false
+    request? : boolean = false
+    apiResponse : any
```
Comment (gwwatkin) [inline ls-ui-react-app/src/appState.ts:12]: See my comment on `apiResponse` below. `any` is generally considered an anti-pattern.

File: ls-ui-react-app/src/pages/UploadCircuitPage.tsx [hunk 4]
```diff
@@ -47,17 +51,28 @@ const UploadACircuit = ( {appState, setAppState} : AppStateProps ) => {
...
+            try {
+                const responseJson = JSON.parse(response.data) as CompilationResult
+                setAppState({
+                    ...appState,
+                    compilationResult: responseJson,
+                })
+            }
+            // catch errors due to parsing JSON
+            catch(err) {
+                // JSON parsing error
```
Comment (gwwatkin) [inline ls-ui-react-app/src/pages/UploadCircuitPage.tsx:68]: Ok I see what you were saying about parse errors, this should set an error somewhere.

Considering the 3 error cases below + this new error case, things might start to get confusing.

To keep track of everything, here is a suggestion:
Make a new datatype that clearly defines the the possible states after a response, maybe something like this:
```ts
type ApiResponse =  null | ServerError | CompilationError | ResponseParseError | CompilationResult
```
where these types are trivial classes like 
```
class ServerError {
    status: number
    errorMsg: string
}

class CompilationError {
   errorMsg: string
}
```
 then you can reduce appState to something super simple like
```ts
class AppState {
    compilationIsLoading : boolean = false
    apiResponse? :  ApiResponse
}
```
And still preserve that nice dispatch of error cases you have below in JSX

**Review by gwwatkin — APPROVED** [review]
Happy to merge this patch. Good improvements on error handling, and I think we are better off having them in sooner than later, so we can work with it. I think we can go deeper about structuring the error cases in a followup PR, added a comment on it.

> The first errors propagate when the API returns nothing, which will generate an error in the AXIOS request if no response is received. This I believe would be a timeout. The errors I get from that are Network Error, and I'm not fully sure where that is originating.

There are two kinds of timeouts.
 * A network timeout: e.g. due to bad wifi, this is a pretty difficult to replicate and rare case - not high priority, can ignore for now.
 * Lambda timing out our code execution. This is going to happen all the time after about 20s of execution. Higher priority, but I think it gets lumped in the 200 errors

> the 2nd case errors are when the API returns a response != 200, such as CORS, or some other HTTP error.

Yep this the trivial kind of error, your handling of it looks good.

> The third case errors are when the API returns a response (200), but encountered some problem with compiling either due to empty file upload or something else. I encountered Type Error: Expected a string or bytes like object.

I see errors like this? 
```
"errorMessage": "Unable to allocate 4.00 GiB for an array with shape (16384, 16384) and data type complex128",
```
Looks like you are handling them fine. The only thing I can suggest is to give more structure to these cases with a datatype. Added a comment.

> Ideally, I would like to only handle errors based on the response object of the AXIOS call, and based on this response, we would know whether JSON.parse(response.data) will succeed.

Yeah that's an ideal case. Practically I think we should still handle `JSON.parse` failing, just to be more robust, to account for programming mistakes, broken configs etc

> There is a subtlety there which we can discuss some time.

Very much so, but it looks you have it all under control. I can only give some strategic advice. If you are free, we can go over it tomorrow.

File: ls-ui-react-app/src/pages/UploadCircuitPage.tsx (hunk 4 above)
Comment (gwwatkin) [reply ls-ui-react-app/src/pages/UploadCircuitPage.tsx:68]: The error/success identification and the `apiResponse` would be set in the `axios` callback

Comment (Keelando) [reply ls-ui-react-app/src/pages/UploadCircuitPage.tsx:68]: Is there a way to pass the AXIOS response object to the .catch block?

Or is there some other way to handle the response object AFTER the JSON.parse error has been caught?

**Comment by Keelando** [comment]
Merging, but will implement suggested changes on next PR

**Pull request merged**